src/GlobalCtxProvider.tsx

```diff
@@ -1,10 +1,83 @@
-import { useMemo } from "react";
-import { type Ctx, GlobalCtx } from "./GlobalCtx";
+import { ApiPromise, WsProvider } from "@polkadot/api";
+import type { TypeRegistry } from "@polkadot/types";
+import { useCallback, useEffect, useMemo, useRef, useState } from "react";
+import { GlobalCtx } from "./GlobalCtx";
+
+export type Status =
+  | { type: "connecting" }
+  | { type: "connected" }
+  | { type: "loading" }
+  | { type: "loaded" }
+  | { type: "failed"; error: string };
+
+export function GlobalCtxProvider({
+  wsAddress,
+  registry,
+  children,
+}: React.PropsWithChildren<{ registry: TypeRegistry; wsAddress: string }>) {
+  const collatorWsRef = useRef<ApiPromise | null>(null);
+  const [status, setStatus] = useState<Status>({ type: "connecting" });
+  const [latestFinalizedBlock, setLatestFinalizedBlock] = useState<number | null>(null);
+
+  const connect = useCallback(async () => {
+    const wsProvider = new WsProvider(wsAddress);
+
+    try {
+      setStatus({ type: "connecting" });
+      console.log(`Connecting to ${wsAddress}`);
+      const polkaStorageApi = await ApiPromise.create({
+        provider: wsProvider,
+        throwOnConnect: true,
+      });
+      collatorWsRef.current = polkaStorageApi;
+
+      await polkaStorageApi.isReady;
+
+      console.log(`Connected to ${await polkaStorageApi.rpc.system.chain()}`);
+
+      const unsub = await collatorWsRef.current.rpc.chain.subscribeFinalizedHeads((header) => {
+        setLatestFinalizedBlock(header.number.toNumber());
+      });
+
+      setStatus({ type: "connected" });
+      return unsub;
+    } catch (error) {
+      console.error(error);
```

Comment:
Won't this print `[object Object]` in certain cases?

Reply:
Not sure honestly, but I'd prefer object object than nothing. 
Previously it gave me nothing with the current error we're having.